File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        md = 0

        # For any given node, the longest path that passes through it
        # is the sum of the height of its left subtree and the height
        # of its right subtree.
        def traversal(root):
            nonlocal md
            if root == None:
                return 0

            if root.left == None and root.right == None:
                return 1

            # Traverse post-order way
            lh = traversal(root.left)
            rh = traversal(root.right)

            # Diameter of the tree is sum of height of left and right subtrees
            md = max(lh + rh, md)

            # What is the height of the subtree?
            # 1 + max(height(l) + height(r))
            return 1 + max(lh, rh)

        traversal(root)
        return md

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool check_valid_bst(TreeNode *root, int min_val, int max_val) {
        if (root == nullptr) {
            return true;
        }

        if (!(min_val < root->val && root->val < max_val)) {
            return false;
        }

        return check_valid_bst(root->left, min_val, root->val) &&
               check_valid_bst(root->right, root->val, max_val);
    }

    bool isValidBST(TreeNode* root) {
        int min_val = INT_MIN;
        int max_val = INT_MAX;
        return check_valid_bst(root, min_val, max_val);
    }
};
